R/capture-requests.R

```diff
@@ -260,24 +260,46 @@ dbGetInfoConTrace <- quote({
   }
 })
 
-dbGetInfoResultTrace <- quote({
-  thing <- returnValue()
-  # TODO: would this be better if we traced the methods individually?
-  if (inherits(dbObj, c("MariaDBResult", "PqResult", "SQLiteResult"))) {
-    hash <- hash(dbObj@sql)
-  } else if (inherits(dbObj, "OdbcResult")) {
-    hash <- hash(dbObj@statement)
+#' Extract a hash from a (result) object
+#'
+#' This function should generally not be used, but must be exported for the
+#' query recording function to work properly
+#'
+#' @param obj the DBI result object to hash
+#'
+#' @return hash for the object
+#' @export
+#' @keywords internal
+hash_db_object <- function(obj) {
+  # TODO: would this be better if we traced the methods using signature?
+  if (inherits(obj, "PostgreSQLResult")) {
+    result_info <- RPostgreSQL::postgresqlResultInfo(obj)
+    hash <- hash(result_info$statement)
+  } else if (inherits(obj, c("MariaDBResult", "PqResult", "SQLiteResult"))) {
+    hash <- hash(obj@sql)
+  } else if (inherits(obj, "OdbcResult")) {
+    hash <- hash(obj@statement)
+  } else if (isS4(obj) && "m_sOperation" %in% methods::slotNames(obj)) {
+      # This is propably a teradata result object, so we can use m_sOperation as hash input.
```

Comment:
I'm not opposed to having this branch for teradata specifically, but curious what your thoughts are on the trade offs between an extra branch here versus having teradata objects use the else with `hash(toString(obj))`. Is there anything that comes through in the `toString(obj)` that isn't helpful for the teradata object? 

Relatedly, thanks for the output of the `str(db_result)`

```
Browse[3]> str(db_result)
Formal class 'TeradataResult' [package "teradatasql"] with 3 slots
..@ m_con :Formal class 'TeradataConnection' [package "teradatasql"] with 7 slots
.. .. ..@ m_uLog :integer64 0
.. .. ..@ m_bImmediate : logi TRUE
.. .. ..@ m_bTraceLog : logi FALSE
.. .. ..@ m_bDebugLog : logi FALSE
.. .. ..@ m_bTimingLog : logi FALSE
.. .. ..@ m_uConnHandle:integer64 2
.. .. ..@ m_envMutable :<environment: 0x000001a65d3313f0>
..@ m_sOperation: 'sql' chr "SELECT * FROM ...."| truncated
..@ m_envMutable:<environment: 0x000001a65a13dcf0>
```

When you do `toString(db_result)` on this same object, I assume (but it would be good to double check) that the two `m_envMutable` slots aren't serialized, is that correct?

Reply:
My thought was, since the teradatasql lib is not present in the CRAN, it is difficult to write a unittest for it. To create and initialize a simple S4 object, I don't need teradatasql lib and can test that part of the code.

As shown above, `toString(obj)` gives out the same as `toString(db_result)'. It was just a different variable in my environment in debugging. Unfortunately I don't know what m_envMutable is and in the toString from TeradataResult I get the values of `uConnHandle=2`, but this is not fixed and therefore a different hash comes out if necessary. I did not like this so much.

For example, since the filename starts with dbGetRowsAffected and the other files use the query for the hash, I wanted to use the query as well. Because of the prefix there are no problems if the hash for Select or dbGetRowsAffected is the same.

Have I answered your open questions? Do you have any further thoughts?

Reply:
>  the toString from TeradataResult I get the values of uConnHandle=2, but this is not fixed and therefore a different hash comes out if necessary. I did not like this so much.

Aaah yes, and this point especially is a good reason for us to have the separate if/else clause here even if it is bespoke to teradata which isn't quite as first-classed as other drivers are.

> My thought was, since the teradatasql lib is not present in the CRAN, it is difficult to write a unittest for it. To create and initialize a simple S4 object, I don't need teradatasql lib and can test that part of the code.

Yes, absolutely your testing approach here is _fantastic;_ thank you for that!